File: dungeons-26.1.2/src/main/java/fishmod/features/other/SearchBar.java
package fishmod.features.other;

import com.mojang.blaze3d.platform.Window;
import fishmod.mixin.accessors.KeyBindingAccessor;
import fishmod.utils.MathParser;
import fishmod.utils.config.values.ExtraOptions;
import fishmod.utils.data.ItemUtil;
import fishmod.utils.rendering.DrawEvents;
import fishmod.utils.rendering.RenderUtils;
import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.Font;
import net.minecraft.client.gui.GuiGraphicsExtractor;
import net.minecraft.client.gui.components.EditBox;
import net.minecraft.client.input.CharacterEvent;
import net.minecraft.client.input.KeyEvent;
import net.minecraft.client.input.MouseButtonEvent;
import net.minecraft.network.chat.Component;
import net.minecraft.world.item.ItemStack;
import org.lwjgl.glfw.GLFW;

public class SearchBar {

    private static final int SEARCH_Y = 20;
    private static final int SEARCH_WIDTH = 150;
    private static final int SEARCH_HEIGHT = 20;
    private static EditBox searchBar;
    private static boolean shouldDisplay = false;
    private static String searchTerm = "";
    private static double parsedValue = Double.NaN;

    public static void init() {
        DrawEvents.INVENTORY_SLOT_AFTER.register((context, item, x, y) -> {
            if (shouldDisplay() && !searchTerm.isEmpty() && ExtraOptions.toggleableSearchBar && Double.isNaN(parsedValue)) {
                if (!matches(item)) {
                    context.fill(x, y, x + 16, y + 16, 0xaa111111);
                }

            }
        });
    }

    private static boolean matches(ItemStack item) {
        String name = item.getHoverName().getString().toLowerCase();
        if (name.equals("air")) return false;

        return (name.contains(searchTerm) || ItemUtil.containsIgnoreCaseLore(item, searchTerm));
    }

    public static void render(GuiGraphicsExtractor context, int mouseX, int mouseY, float deltaTicks) {
        if (!exists() || !shouldDisplay() || !ExtraOptions.toggleableSearchBar) return;
        searchBar.extractRenderState(context, mouseX, mouseY, deltaTicks);

        if (!Double.isNaN(parsedValue)) {
            String expression = "  §e= §2" + RenderUtils.formatNumber((float) parsedValue);
            Font textRenderer = Minecraft.getInstance().font;
            if (textRenderer == null) return;

            int textX = searchBar.getX() + textRenderer.width(searchTerm) + 4;
            int textY = searchBar.getY() + (searchBar.getHeight() - 8) / 2;
            context.text(textRenderer, expression, textX, textY, 0xffffffff, true);
        }
    }

    public static boolean keyPressed(KeyEvent input) {
        if (!exists() || !ExtraOptions.toggleableSearchBar) return false;

        boolean ctrlIsPressed = (input.modifiers() & GLFW.GLFW_MOD_CONTROL) != 0;

        if (ctrlIsPressed && input.key() == GLFW.GLFW_KEY_F) {
            shouldDisplay = !shouldDisplay;
            return true;
        } else if (shouldDisplay() && searchBar.isFocused()) {
            // Never eat the player's drop key — pressing it should drop the item, not type into search.
            // Unfocus the search so the keystroke falls through to vanilla's drop handling.
            try {
                Minecraft mc = Minecraft.getInstance();
                int dropCode = ((KeyBindingAccessor) (Object) mc.options.keyDrop).getBoundKey().getValue();
                if (input.key() == dropCode) { searchBar.setFocused(false); return false; }
            } catch (Exception ignored) {}
            if (input.key() == GLFW.GLFW_KEY_ENTER) {
                if (!Double.isNaN(parsedValue)) {
                    searchBar.setValue(RenderUtils.formatNumber((float) parsedValue));
                }
            } else if (input.key() != GLFW.GLFW_KEY_ESCAPE) {
                searchBar.keyPressed(input);
                return true;
            }
        }
        return false;
    }

    public static void CharTyped(CharacterEvent input) {
        if (!exists() || !searchBar.isFocused() || !shouldDisplay() || !ExtraOptions.toggleableSearchBar) return;
        searchBar.charTyped(input);
    }

    public static void onMouseClick(MouseButtonEvent click) {
        if (!exists() || !shouldDisplay() || !ExtraOptions.toggleableSearchBar) return;
        searchBar.setFocused(inBounds(click.x(), click.y()));
    }


    public static boolean shouldDisplay() {
        return shouldDisplay;
    }

    private static boolean exists() {
        if (searchBar != null) return true;

        Minecraft mc = Minecraft.getInstance();
        Font textRenderer = mc.font;
        Window window = mc.getWindow();

        if (window == null || textRenderer == null) return false;

        searchBar = new EditBox(textRenderer, (window.getGuiScaledWidth() - SEARCH_WIDTH) / 2, SEARCH_Y, SEARCH_WIDTH, SEARCH_HEIGHT, Component.literal(""));
        searchBar.setResponder(string -> {
            searchTerm = string.toLowerCase();
            parsedValue = MathParser.parseExpression(searchTerm);
        });
        return true;
    }

    private static boolean inBounds(double x, double y) {
        int sx = searchBar.getX();
        int sy = searchBar.getY();
        int sw = searchBar.getWidth();
        int sh = searchBar.getHeight();

        return x >= sx && x <= sx + sw && y >= sy && y <= sy + sh;
    }

}

File: dungeons-26.1.2/src/main/kotlin/fishmod/features/other/SearchBar.kt
package fishmod.features.other

import com.mojang.blaze3d.platform.Window
import fishmod.mixin.accessors.KeyBindingAccessor
import fishmod.utils.MathParser
import fishmod.utils.config.values.FishSettings
import fishmod.utils.data.ItemUtil
import fishmod.utils.rendering.DrawEvents
import fishmod.utils.rendering.RenderUtils
import net.minecraft.client.Minecraft
import net.minecraft.client.gui.Font
import net.minecraft.client.gui.GuiGraphicsExtractor
import net.minecraft.client.gui.components.EditBox
import net.minecraft.client.input.CharacterEvent
import net.minecraft.client.input.KeyEvent
import net.minecraft.client.input.MouseButtonEvent
import net.minecraft.network.chat.Component
import net.minecraft.world.item.ItemStack
import org.lwjgl.glfw.GLFW

object SearchBar {

    private const val SEARCH_Y = 20
    private const val SEARCH_WIDTH = 150
    private const val SEARCH_HEIGHT = 20
    private var searchBar: EditBox? = null
    private var shouldDisplayVal = false
    private var searchTerm = ""
    private var parsedValue = Double.NaN

    @JvmStatic
    fun init() {
        DrawEvents.INVENTORY_SLOT_AFTER.register { context, item, x, y ->
            if (!FishSettings.inventorySearchEnabled || !shouldDisplay() || searchTerm.isEmpty() || !parsedValue.isNaN()) return@register
            if (!matches(item)) {
                context.fill(x, y, x + 16, y + 16, 0xaa111111.toInt())
            } else if (FishSettings.inventorySearchHighlight) {
                val c = FishSettings.inventorySearchHighlightColor
                context.fill(x, y, x + 16, y + 1, c)
                context.fill(x, y + 15, x + 16, y + 16, c)
                context.fill(x, y + 1, x + 1, y + 15, c)
                context.fill(x + 15, y + 1, x + 16, y + 15, c)
            }
        }
    }

    private fun matches(item: ItemStack): Boolean {
        val name = item.hoverName.string.lowercase()
        if (item.isEmpty || name == "air") return false

        return name.contains(searchTerm) || ItemUtil.containsIgnoreCaseLore(item, searchTerm)
    }

    @JvmStatic
    fun render(context: GuiGraphicsExtractor, mouseX: Int, mouseY: Int, deltaTicks: Float) {
        if (!FishSettings.inventorySearchEnabled || !exists() || !shouldDisplay()) return
        val bar = searchBar!!
        // re-centre every frame so a window resize doesn't strand the bar
        bar.x = (Minecraft.getInstance().window.guiScaledWidth - SEARCH_WIDTH) / 2
        bar.extractRenderState(context, mouseX, mouseY, deltaTicks)

        if (!parsedValue.isNaN()) {
            val expression = "  §e= §2" + RenderUtils.formatNumber(parsedValue.toFloat())
            val textRenderer: Font = Minecraft.getInstance().font ?: return

            val textX = bar.x + textRenderer.width(searchTerm) + 4
            val textY = bar.y + (bar.height - 8) / 2
            context.text(textRenderer, expression, textX, textY, 0xffffffff.toInt(), true)
        }
    }

    @JvmStatic
    fun keyPressed(input: KeyEvent): Boolean {
        if (!FishSettings.inventorySearchEnabled || !exists()) return false
        val bar = searchBar!!

        val ctrlIsPressed = (input.modifiers() and GLFW.GLFW_MOD_CONTROL) != 0

        if (ctrlIsPressed && input.key() == GLFW.GLFW_KEY_F) {
            if (FishSettings.inventorySearchAlwaysShow) {
                bar.isFocused = !bar.isFocused
            } else {
                shouldDisplayVal = !shouldDisplayVal
                bar.isFocused = shouldDisplayVal
            }
            return true
        } else if (shouldDisplay() && bar.isFocused) {
            try {
                val mc = Minecraft.getInstance()
                val dropCode = (mc.options.keyDrop as KeyBindingAccessor).boundKey.value
                if (input.key() == dropCode) { bar.isFocused = false; return false }
            } catch (ignored: Exception) {
            }
            if (input.key() == GLFW.GLFW_KEY_ENTER) {
                if (!parsedValue.isNaN()) {
                    bar.value = RenderUtils.formatNumber(parsedValue.toFloat())
                }
            } else if (input.key() != GLFW.GLFW_KEY_ESCAPE) {
                bar.keyPressed(input)
                return true
            }
        }
        return false
    }

    @JvmStatic
    fun CharTyped(input: CharacterEvent) {
        if (!FishSettings.inventorySearchEnabled || !exists() || !searchBar!!.isFocused || !shouldDisplay()) return
        searchBar!!.charTyped(input)
    }

    @JvmStatic
    fun onMouseClick(click: MouseButtonEvent) {
        if (!FishSettings.inventorySearchEnabled || !exists() || !shouldDisplay()) return
        searchBar!!.isFocused = inBounds(click.x(), click.y())
    }

    @JvmStatic
    fun shouldDisplay(): Boolean = shouldDisplayVal || FishSettings.inventorySearchAlwaysShow

    private fun exists(): Boolean {
        if (searchBar != null) return true

        val mc = Minecraft.getInstance()
        val textRenderer: Font? = mc.font
        val window: Window? = mc.window

        if (window == null || textRenderer == null) return false

        val bar = EditBox(
            textRenderer, (window.guiScaledWidth - SEARCH_WIDTH) / 2, SEARCH_Y, SEARCH_WIDTH, SEARCH_HEIGHT, Component.literal("")
        )
        bar.setResponder { string ->
            searchTerm = string.lowercase()
            parsedValue = MathParser.parseExpression(searchTerm)
        }
        searchBar = bar
        return true
    }

    private fun inBounds(x: Double, y: Double): Boolean {
        val bar = searchBar!!
        val sx = bar.x
        val sy = bar.y
        val sw = bar.width
        val sh = bar.height

        return x >= sx && x <= sx + sw && y >= sy && y <= sy + sh
    }
}
